### PR TITLE
Detect cycles during type instantiation

### LINKED ISSUE #3237: Stack Overflow with InstatiateType
```typescript
function foo<T>(x: T) {
    var y: {
        y1: T;
        y2: typeof y
    };
    return y;
}


function bar<T>(x: T) {
    var y: {
        y1: T;
        y2: typeof y;
    }
    return y;
}

var a = foo<string>("hello");
var b = bar<number>(5);
a = b;
```
stack_overflow_instantiatetype

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #3237.

### CODE REVIEW

**Comment by mhegazy** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/compiler/types.ts [hunk 1]
```diff
@@ -1589,6 +1589,7 @@ module ts {
     /* @internal */
     export interface TypeMapper {
         (t: TypeParameter): Type;
+        mappings?: Map<Type>;  // Type mapping cache
```
Comment (JsonFreeman) [inline src/compiler/types.ts]: Can you write what the key and the value represent in the map?

**Comment by JsonFreeman** [comment]
:+1:

**Pull request merged**